File: dinosay.ts
function getFlags(args: string[]) {
    const flags = {
        deno: false,
        trex: false,
        help: false,
        version: false,
    };
    if (args.length == 0) {
        flags.help = true;
    }
    for (let i = 0; i < args.length; i++) {
        if (args[i] == "--help" || args[i] == "-h") {
            flags.help = true;
        } else if (args[i] == "--version" || args[i] == "-v") {
            flags.version = true;
        } else if (args[i] == "--deno" || args[i] == "-d") {
            flags.deno = true;
        } else if (args[i] == "--trex" || args[i] == "-t") {
            flags.trex = true;
        }
    }
    return flags;
}

function removeFlags(args: string[]) {
    const newArgs: string[] = [];
    for (let i = 0; i < args.length; i++) {
        if (args[i] != "--help" && args[i] != "-h" && args[i] != "--version" && args[i] != "-v" && args[i] != "--deno" && args[i] != "-d" && args[i] != "--trex" && args[i] != "-t") {
            newArgs.push(args[i]);
        }
    }
    return newArgs;
}


function lineLongerThanBestSoFar(len: number) {
    if (len > maxWidthSeen)
        maxWidthSeen = len;
}

const flags = getFlags(Deno.args);
if (flags.help) {
    console.log("Usage: dinosay [options] [text]");
    console.log("Options:");
    console.log("  -h, --help     Show this help message");
    console.log("  -v, --version  Show version information");
    console.log("  -d, --deno     Use the Deno dinosaur");
    console.log("  -t, --trex     Use the T-Rex dinosaur");
    Deno.exit(0);
}
if (flags.version) {
    console.log("dinosay v1.0.0");
    Deno.exit(0);
}

const args = removeFlags(Deno.args);


const MAX_WIDTH = 30;
let maxWidthSeen = 0;
let lines: string[] = [];
let index = 0;
let line = "";

while (index < args.length) {
    if (line.length + args[index].length > MAX_WIDTH) {
        lines.push(line)
        lineLongerThanBestSoFar(line.length);
        line = ""
    }
    line += " " + args[index]
    if (index == args.length - 1) {
        lines.push(line)
    }
    lineLongerThanBestSoFar(line.length);
    index++;
}

let bubble = "        __"
for (let i = 0; i < maxWidthSeen; i++) {
    bubble += "_";
}
bubble += "\n       /  "
for (let i = 0; i < maxWidthSeen; i++) {
    bubble += " "
}
bubble += "\\";

// Actually print the text
for (const line of lines) {
    bubble += "\n"
    bubble += "      |  "
    bubble += line
    if (line.length != maxWidthSeen) {
        const difference = maxWidthSeen - line.length;
        for (let i = 0; i < difference; i++) {
            bubble += " ";
        }
    }
    bubble += "  |"
}

// Close the bubble
bubble += "\n       \\  "
for (let i = 0; i < maxWidthSeen; i++) {
    bubble += "_";
}
bubble += "/"


const trex = `
        \\/
            ___________
           /' O   O    \\
           \\            \\
            \`^^^^^\\      \\                          ___
                  /       \\                        {_  \\
             /^^^^^        \\                         \\  |
             \\_____         \\                         | |
                    \\        \\_________            __/ /
                     |         ___     -----_____--   /
                   >-|        /   \\           _____--\`
                    >-\\      (     ) _____-----
                       \\______\\    \\-
                           \\   \\    \\
                            \\   \\    \\
                            /   /    /
                          _/  _/    /
                         /___/_____/
			`
const deno = `
        \\/
       __
      / •) 
     /  /    ____
     \\  \\_-\`\`    \`\`-_     
      \\              \`-_  ,
       \`-,_   ___  ___--\`\`
          ||_|  ||_|                
`
let final = bubble;
if (flags.deno) {
    final += deno;
} else {
    final += trex;
}
console.log(final)

File: dinosay.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

struct Flags {
    bool deno = false;
    bool trex = false;
    bool help = false;
    bool version = false;
};

Flags getFlags(const std::vector<std::string>& args) {
    Flags flags;
    if (args.empty()) {
        flags.help = true;
    }
    for (const std::string& arg : args) {
        if (arg == "--help" || arg == "-h") {
            flags.help = true;
        } else if (arg == "--version" || arg == "-v") {
            flags.version = true;
        } else if (arg == "--deno" || arg == "-d") {
            flags.deno = true;
        } else if (arg == "--trex" || arg == "-t") {
            flags.trex = true;
        }
    }
    return flags;
}

std::vector<std::string> removeFlags(const std::vector<std::string>& args) {
    std::vector<std::string> newArgs;
    for (const std::string& arg : args) {
        if (arg != "--help" && arg != "-h" && arg != "--version" && arg != "-v" && arg != "--deno" && arg != "-d" && arg != "--trex" && arg != "-t") {
            newArgs.push_back(arg);
        }
    }
    return newArgs;
}

int maxWidthSeen = 0;

void lineLongerThanBestSoFar(int len) {
    if (len > maxWidthSeen) {
        maxWidthSeen = len;
    }
}

int main(int argc, char* argv[]) {
    std::vector<std::string> args(argv + 1, argv + argc);
    Flags flags = getFlags(args);

    if (flags.help) {
        std::cout << "Usage: dinosay [options] [text]" << std::endl;
        std::cout << "Options:" << std::endl;
        std::cout << "  -h, --help     Show this help message" << std::endl;
        std::cout << "  -v, --version  Show version information" << std::endl;
        std::cout << "  -d, --deno     Use the Deno dinosaur" << std::endl;
        std::cout << "  -t, --trex     Use the T-Rex dinosaur" << std::endl;
        return 0;
    }
    if (flags.version) {
        std::cout << "dinosay v1.0.0" << std::endl;
        return 0;
    }

    std::vector<std::string> textArgs = removeFlags(args);

    const int MAX_WIDTH = 30;
    std::vector<std::string> lines;
    int index = 0;
    std::string line = "";

    while (index < textArgs.size()) {
        if (line.length() + textArgs[index].length() + (line.empty() ? 0 : 1) > MAX_WIDTH) {
            lines.push_back(line);
            lineLongerThanBestSoFar(line.length());
            line = "";
        }
        if (!line.empty()) {
            line += " ";
        }
        line += textArgs[index];
        if (index == textArgs.size() - 1) {
            lines.push_back(line);
        }
        lineLongerThanBestSoFar(line.length());
        index++;
    }

    std::string bubble = "          __";
    for (int i = 0; i < maxWidthSeen; ++i) {
        bubble += "_";
    }
    bubble += "\n         /  ";
    for (int i = 0; i < maxWidthSeen; ++i) {
        bubble += " ";
    }
    bubble += "\\";

    for (const std::string& currentLine : lines) {
        bubble += "\n        |  ";
        bubble += currentLine;
        if (currentLine.length() != maxWidthSeen) {
            int difference = maxWidthSeen - currentLine.length();
            for (int i = 0; i < difference; ++i) {
                bubble += " ";
            }
        }
        bubble += "  |";
    }

    bubble += "\n         \\__";
    for (int i = 0; i < maxWidthSeen; ++i) {
        bubble += "_";
    }
    bubble += "/";

    const std::string trex = R"(
            \/
            ___________
           /' O   O    \
           \            \
            `^^^^^^\     \
                   /      \                       <==~,
             /^^^^^        \                         \ \
             \______        \                         | |
                    \        \_________              /  |
                     |         ____    ``----_____--`  /
                   >-|        /    \           _____--` 
                    >-\      (      )______-----
                       \______\     \
                           \   \     \
                            \   \     \
                            /   /     /
                          _/  _/     /
                         /___/______/
)";
    const std::string deno = R"(
            \/
       __
      / •) 
     /  /    ____
     \  \_-``    ``-_     
      \              `-_  ,
       `-,_   ___  ___--``
          ||_|  ||_|  
)";

    std::string final = bubble;
    if (flags.deno) {
        final += deno;
    } else {
        final += trex;
    }
    std::cout << final << std::endl;

    return 0;
}